Add rust friendly `assign parameters` methods

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Follow up on #12794.
These commits expose some rust-friendly `assign_parameters` methods in `CircuitData` allowing us to perform parameter assignments without having call to python directly.


### Details and comments
The following commits add the methods `assign_paramters_from_slice` and `assign_parameters_from_mapping` which enable the assignment of parameters in rust in a friendly way instead of using hard-Python structures (`Py<_>` or `Bound<'_, _>`. Here are their use cases:
- `assign_paramters_from_slice`: Use whenever a collection of `Param` needs to be assigned. The amount of parameters sent needs to be of the same size as the the number of parameters in the `CircuitData` instance.
- `assign_parameters_from_mapping`: Use whenever of mapping between `ParameterUuid` instances from the `CircuitData` and `Param` instances coming from `Gates` or `Operations`. This can be used for partial assignment, and the number of parameters can be smaller or equal to the amount of parameters in the `CircuitData` instance.

### Known issues
- ~~The mapping of `Params` : `Params` may not be optimal as we wouldn't know if the types of parameters we're binding are valid, `Param::Float` does not posses a `uuid` and therefore is not a valid key for the mapping.~~ Will do a mapping between `ParameterUuid` and `Param` to assume that all keys are valid instances with valid `uuid`s.
- ~~The mapping of `assign_parameters_from_mapping` does not accept instances of `Map<Param, Param>` due to param not implementing the hash property in Rust.~~ Now it accepts any mapping of `ParameterUuid` : `Param` as long as it can be turned into an iterator of type `(ParameterUuid, Param)`.

### Blockers
- [x] #12926

### Other comments
- Feel free to point out any issues in the comments and I will quickly address.

raynelfss committed:
Initial: Add `assign_parameter` methods that can be used from Rust.

raynelfss committed:
Merge branch 'Qiskit:main' into add-rust-bind-params

raynelfss committed:
Add: Error message in `from_slice` method.
- Accept owned mapping for `from_mapping`.

raynelfss committed:
Fix: Use `assign_parameters_from_slice` as inner method for `assign_parameters_sequence`

raynelfss committed:
Merge branch 'main' into add-rust-bind-params

raynelfss committed:
Merge branch 'main' into add-rust-bind-params

raynelfss committed:
Fix: Use `ParameterUuid` as keys for `assign_parameter_from_mapping`

raynelfss committed:
Fix: Accept Param references in `assign_parameters_inner`

raynelfss committed:
Merge branch 'main' into add-rust-bind-params

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -949,45 +949,42 @@ impl CircuitData {
                     .iter()
                     .zip(old_table.drain_ordered())
-                    .map(|(value, (param_ob, uses))| (param_ob, Param::Float(*value), uses)),
+                    .map(|(value, (obj, uses))| (obj, value, uses)),
```
It's not clear to me that an entire reallocation of the entire array here is suitable in order to accept `&'a Param` in the inner methods.  It suggests to me that there's ownership problems going on, or that the new interface isn't entirely correct - we didn't need to allocate before, but we are now, so there's a regression.

I suspect that there's a way to make this work without allocation even in the current signature of the inner method, but if not, two points:
1. `Param` is very cheap to clone if the GIL is held; _mostly_ it's just going to be copying floats, which are `Copy`, and the rest of it is going to be ref-count updates in Python space.
2. We might be able to change the signature of the inner assigners to accept the iterator item as `AsRef<Param>`, and define `AsRef<Param> for Param` (there's no reflexive blanket impl built into the language only as a technical limitation).  That would let the inner method handle both `Param` and `&Param` iterators monomorphically.

raynelfss replied:
**1.**  Would it be possible for me to clone `Param` safely by doing something like this:
```rs
...
match param {
    Param::Float(float) => Param::Float(*float),
    Param::ParameterExpression(exp) => Param::ParameterExpression(exp.clone_ref(py)),
    Param::Object(obj) => Param::Object(obj.clone_ref(py)),
}
...
```
If so, maybe this should be a method added to `Param`, if it is not as expensive to simply call `.clone()` considering that this will most likely copy floats usually.

**2.** So if we're able to implement the `AsRef<Param>` trait, this would allow us to receive both referenced and owned `Param instances as long as we specify it? Hmm... I'll give it a try then.

jakelishman replied:
1. yes it's fine to add that, and for PyO3 style, we'd probably call that method `clone_ref`
2. well, the method would be generic and the Rust compiler actually ends up compiling two versions of it (one for the reference form, one for the owned form), but in terms of _writing_ the code, it appears like we just wrote one version

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -949,45 +949,42 @@ impl CircuitData {
-            items.push((param_ob, value.0, self.param_table.pop(uuid)?));
+            items.push(value);
+            objs.push(param_ob); // We need to separate the objects to avoid cloning.
         }
```
Can you explain why separating into multiple collections avoids cloning?  It feels like it shouldn't matter, or at a minimum all we'd be cloning is Python-space pointers anyway, which are very cheap to clone.

raynelfss replied:
It was an incorrect way of passing `Param` as a reference. However, I will remove this for a better structure.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1135,6 +1132,44 @@ impl CircuitData {
+                param_obj.unwrap().clone_ref(py),
+                value,
+                self.param_table.pop(param_uuid)?,
+            ));
```
Since we're already returning a `PyResult` from this function, I don't think there's any pain from having it correctly handle the error condition of a bad `ParameterUuid` here.

jakelishman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -1135,6 +1132,44 @@ impl CircuitData {
+        )
+    }
+
+    /// Assigns parameters to circuit data based on a mapping of `ParameterUuid` : `Param`.
+    /// This mapping assumes that the provided `ParameterUuid` keys are instances
+    /// of `ParameterExpression`.
```
It's conventional in Rust documentation to put the assumptions that cause panics in a section called `# Panics` to really make them stand out - idiomatic Rust is _far_ stricter about errors than Python, where the conventions are more like "just raise an exception".  Panics are serious business.

(But also see the other comment - imo it'd be cleaner just to handle the error properly, since it won't cost us anything.)

raynelfss replied:
Sounds good!

jakelishman reviewed (commented):
Thanks for this Ray.  My biggest concern here is in the first review comment - the new interface gives the Python-space forms additional allocations that they didn't need before.  I think we ought to either rework the Python forms (which I suspect might be possible) or at least rework the inner interface a little such that those allocations aren't necessary; they logically aren't, since they weren't there before.

raynelfss committed:
Fix: Ownership issues
- Remove initial allocation of `Param` instances within a vec in `assign_parameters_iterable`.
- Revert changes in `assign_parameters_mapping`.
- Fix error message in `assign_parameter_from_slice`.
- Return an error if a `uuid` is not found in `assign_parameters from mapping. Also, return error if a valid uuid is not found to avoid panicking.
- Add `clone_ref(py)` method to `Param` to ensure that we use the most efficient cloning methods (either `clone_ref` with gil or copying).
- Implement trait `AsRef<Param>` for `Param` to be able to send both owned and non-owned instances to `assign_parameters_inner`.

jakelishman commented on crates/circuit/src/operations.rs:
```diff
@@ -125,6 +125,21 @@ impl Param {
+    }
+}
+
+impl AsRef<Param> for Param {
+    fn as_ref(&self) -> &Param {
+        self
+    }
```
Might be worth just putting in a comment here about why this exists potentially with a link to the Rust docs bit about why the blanket impl doesn't: https://doc.rust-lang.org/std/convert/trait.AsRef.html#reflexivity

jakelishman reviewed (commented):
Thanks Ray, this looks much cleaner to me now, and it's good to see no new allocations added in the existing paths.  One tiny comment about a comment is all, but code-wise this is good to merge.

raynelfss committed:
Docs: Add comment on `AsRef<Param>` impl block.

jakelishman reviewed (approved):
Ace, thanks!
